### PR TITLE
Warns the user for deprecated flags that use _ separators

### PR DESCRIPTION
**Pull request by andronat**
Fix for #7478

Also gets rid of "intermediate" representation of flags with `.` to enforce consistent flag name format throughout the project. 

i.e. Formats like `flag_f`, `flag-f` and `flag.f` were all valid in the source code, now we throw a warning if a developer uses the format of `flag_f` and `flag.f` is invalid.

### CODE REVIEW

**Comment by andronat** [comment]
I signed it!

**Comment by jlowdermilk** [comment]
Thanks for the PR!

LGTM, but I know @smarterclayton has expressed distaste with glog, so I'll let him comment.

**Comment by smarterclayton** [comment]
It's fine for now - we need to go through and figure out what our log story is for all CLI commands anyway so this doesn't make it any worse.

**Comment by andronat** [comment]
Don't merge yet, sorry but I just found that it is not working with all the flags.

And I also have a situation that I am not sure on how to handle it. There are flags that are coming from other packages that are written with `_`. e.g. `log_dir` and `log_backtrace_at` from the `glog` package. How am I suppose to handle them? Sending a depreciation warning I think is useless on the situation.

**Comment by jlowdermilk** [comment]
I agree a deprecation warning on flags merged from other packages doesn't make sense. How we handle depends on whether we plan to continue using glog in kubectl or not. If yes, then we can't avoid having a mix of `_` and `-` flags. @ghodss, thoughts?

**Comment by andronat** [comment]
Maybe we can "shadow" them? spf13/pflag#23

**Comment by jlowdermilk** [comment]
I think a better path is to do the translation ourselves when we import the flags into the kubectl command flagset [here](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/kubectl/cmd/util/factory.go#L222); see also [AddFlagSetToPFlagSet](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/pkg/util/pflag_import.go#L89) . Doing the translation inside pflag seems wrong to me.

**Comment by ghodss** [comment]
@jlowdermilk's solution sounds great to me. The alternative is keeping a whitelist inside WordSepNormalizeFunc, which sounds worse.

**Comment by andronat** [comment]
Ok, so I will translate all the persistent flags and then I will enable the `_` warning messages

**Comment by andronat** [comment]
spf13/pflag#23 and spf13/cobra#110 are merged please review.

**Comment by jlowdermilk** [comment]
Looks almost ready. Can you rebase to 2 commits (1 to bump cobra/pflag, 1 with your changes)? Also, you'll need to run `hack/run-gendocs.sh` as the auto-generated `kubectl` docs have changed now that inherited flags are `-` separated. You will actually get a presubmit check telling to run gendocs, if you install the [hooks](https://github.com/GoogleCloudPlatform/kubernetes/blob/master/docs/devel/development.md#hooks).

**Comment by andronat** [comment]
Done, sorry for the docs.

**Comment by jlowdermilk** [comment]
The cobra bump picked up https://github.com/spf13/cobra/pull/109, which is causing the default GenMarkdownTree to prepend the filename to each generated md file, which we do not want.

If you don't mind waiting a bit for https://github.com/spf13/cobra/pull/112 to be merged, and re-updating cobra that would be simplest. Otherwise, quick fix is to replace [this line](https://github.com/andronat/kubernetes/blob/fix_7478/cmd/gendocs/gen_kubectl_docs.go#L51) with
```
GenMarkdownTreeCustom(kubectl, outDir, func(s string) string { return "" }, func(s string) string { return s })
```
and leave a todo to update once https://github.com/spf13/cobra/pull/112 is in.

**Comment by eparis** [comment]
@jlowdermilk merged fix in cobra.

**Comment by jlowdermilk** [comment]
Thanks @eparis! @andronat, if you update cobra and rerun `hack/run-gendocs.sh` tests should be passing.

File: pkg/util/pflag_import.go [hunk 1]
```diff
@@ -85,23 +85,25 @@ func addFlagToPFlagSet(f *flag.Flag, fs *pflag.FlagSet) {
...
 	fsIn.VisitAll(func(f *flag.Flag) {
 		addFlagToPFlagSet(f, fsOut)
 	})
 }
 
-// Add all of the top level 'flag' package flags to the top level 'pflag' flags.
+// AddAllFlagsToPFlags adds all of the top level 'flag' package flags to the top level 'pflag' flags.
 func AddAllFlagsToPFlags() {
 	AddFlagSetToPFlagSet(flag.CommandLine, pflag.CommandLine)
 }
```
Comment (eparis) [inline pkg/util/pflag_import.go]: Is this even needed, after https://github.com/spf13/cobra/pull/75

I know it's not the point of your PR, but how much of pflag_import.go can we rip out?

Comment (eparis) [inline pkg/util/pflag_import.go]: Something still isn't quite right around this stuff.  ignore me.

Comment (andronat) [inline pkg/util/pflag_import.go]: Actually I was also trying to get rid of those funcs there. The problem is that cobra is merging flag command line flags but is not working for the pflag global flags. Also an other problem is the order and "when" each of those merges are taking place. This is actually an issue of cobra and I might work on that at some point: spf13/cobra#44

**Comment by jlowdermilk** [comment]
lgtm

File: pkg/util/pflag_import.go (hunk 1 above)
Comment (eparis) [inline pkg/util/pflag_import.go]: Agree, we should work on this in cobra.

**Comment by andronat** [comment]
Rebased to resolve conflicts

**Comment by eparis** [comment]
LGTM

**Comment by dchen1107** [comment]
You have to rebase, sorry for this.

**Comment by andronat** [comment]
Done

**Comment by andronat** [comment]
any news on that? It is ready to merge, I rebased it to latest master

**Comment by eparis** [comment]
I haven't tested it, but does using --flag_name= on the command line give a warning, or does it just accept things silently?

**Comment by andronat** [comment]
It is like that:
```
kubectl --no_headers=true get cs
W0527 14:31:51.039177     373 flags.go:36] no_headers is DEPRECATED and will be removed in a future version. Use no-headers instead.
controller-manager   Healthy   ok        nil
scheduler            Healthy   ok        nil
```

**Comment by thockin** [comment]
Are you guys ready to go with this - it is a substantial change, so I want to get it off the books one way or another.

**Comment by eparis** [comment]
LGTM

**Comment by eparis** [comment]
(but it's not assigned to me, so I'm not adding the label)

**Comment by jlowdermilk** [comment]
Still LGTM

**Comment by thockin** [comment]
fire in the  hole

**Pull request merged**